enhancements/multi-arch/dynamic-imagestream-importmode-setting.md

```diff
@@ -0,0 +1,390 @@
+
+## Proposal
+
+The proposal is for CVO to expose the
+architecture/type of payload installed on the
+cluster and for apiserver to use that information
+to set the importMode for imagestreams globally.
+
+### Workflow Description
+ 
+#### Installation
+1. User triggers installation through
+openshift-installer
+2. CVO inspects the payload and updates the status
+field of `ClusterVersion` with the payload
+architecture/type (inferred through
+`release.openshift.io/architecture`)
+3. openshift-apiserver operator looks to see if
+the image config CRD has the
+`ImagestreamImportMode` set. If it does, get that
+value, if not look at the `ClusterVersion` status
+and gets the value of the payload type.
+4. the operator updates the `OpenshiftAPIServer`'s
+observed config ImagePolicyConfig field with the
+import mode based on the status.
+5. openshift-apiserver uses the value of the
+import mode in the observed config's
+ImagePolicyConfig field to set the default import
+mode for any newly created imagestream.
+
+#### Upgrade
+
+##### Case-1: upgrade when no import mode is
+specified in `image.config.openshift.io`
+1. User has a cluster installed with a single arch
+payload
+2. User triggers an update to multi-arch payload
+using `oc adm upgrade --to-multi-arch`
+3. CVO triggers the update, inspects the payload
+and updates the status field of `ClusterVersion`
+with the payload architecture/type
+4. The rest is the same as steps 3-5 in the
+install section
+
+##### Case-2: upgrade when import mode is
+specified in `image.config.openshift.io`
+1. User has a cluster installed with a single arch
+payload
+2. User sets the import mode of the image config
+CRD's `spec.ImagestreamImportMode` field.
+3. User triggers an update to multi-arch payload
+using `oc adm upgrade --to-multi-arch`
+4. CVO triggers the update, inspects the payload
+and updates the status field of `ClusterVersion`
+with the payload architecture/type
+5. openshift-apiserver operator looks to see if
+the image config CRD has the
+`ImagestreamImportMode` set. Since it has been
+set, it preserves that value in the observed
+config and does not change it based on the payload
+type advertised by CVO.
+
+#### Manual
+1. User updates the `ImagestreamImportMode` spec
+field in the `image.config.openshift.io` `cluster`
+CR.
+3. openshift-apiserver operator reconciles the CR
+to see if the `ImagestreamImportMode` value is set
+and gets that value.
+4. the operator updates the `OpenshiftAPIServer`'s
+observed config ImagePolicyConfig field with the
+import mode.
+5. openshift-apiserver uses the value of the
+import mode in the observed config's
+ImagePolicyConfig field to set the default import
+mode for any newly created imagestream.
+
+Note that a change to the import mode in the image
+config CR would trigger a redeployment of the
+openshift-apiserver
+
+### API Extensions
+
+- image.config.openshift.io: Add
+  `ImagestreamImportMode` config flag to CRD spec.
+
+``` 
+type ImageSpec struct {
```

Comment:
need a status field as well. The imageregistry operator will set the status based on .spec+CVO status.  The openshift-apiserver-operator will only consume the status and will require the status prior to starting the first time.

Reply:
Is there an issue if the openshift-apiserver-operator directly consumes the ClusterVersion status?

Reply:
I guess this makes sense since the status field of `images.config.openshift.io` is the only field the operator would need to look at to make the decision and that field would be updated by the imageregistry operator based on the ClusterVersion status/ the imagestreamImportMode field in the spec

Reply:
made the appropriate changes to based on adding a field in the status.